File: debug/src/main/java/com/sofar/debug/performance/monitor/AbsMonitor.java
package com.sofar.debug.performance.monitor;

import android.app.Application;
import android.os.Handler;
import androidx.annotation.NonNull;

import com.sofar.debug.performance.PerformanceMonitorManager;

public abstract class AbsMonitor implements IMonitor {

  protected boolean started;

  @NonNull
  protected Handler handler;
  @NonNull
  protected Handler UIHandler;
  @NonNull
  protected Application appContext;

  public AbsMonitor() {
    handler = PerformanceMonitorManager.get().getHandler();
    UIHandler = PerformanceMonitorManager.get().getUIHandler();
    appContext = PerformanceMonitorManager.get().getAppContext();
  }

  @Override
  public void start() {
    if (started) {
      return;
    }

    started = true;
    onStart();
  }

  @Override
  public void stop() {
    if (!started) {
      return;
    }

    started = false;
    onStop();
  }

  protected abstract void onStart();

  protected abstract void onStop();
}

File: tool/profiler/src/main/java/com/sofar/profiler/AbsMonitor.kt
package com.sofar.profiler

import android.app.Application
import android.os.Handler

abstract class AbsMonitor : IMonitor {

  protected var started = false

  protected var handler: Handler = MonitorManager.monitorHandler
  protected var UIHandler: Handler = MonitorManager.UIHandler
  protected var appContext: Application = MonitorManager.appContext
  protected var pid = MonitorManager.pid.toString()


  override fun start() {
    if (started) {
      return
    }
    started = true
    onStart()
  }

  override fun stop() {
    if (!started) {
      return
    }
    started = false
    onStop()
  }

  protected abstract fun onStart()

  protected abstract fun onStop()
}